design/decisions/0006-tedge-flows.md

```diff
@@ -0,0 +1,157 @@
+  - The function is expected to return zero, one or many transformed messages `[{ topic: string, payload: string }]`
+  - An exception can be thrown if the input message cannot be transformed.
+- A flow script can also export an `update_config` function
+  - This function is called on each message received on the `meta_topics` as defined in the config.
+  - The arguments are:
+    - The message to be interpreted as a config update `{ topic: string, payload: string }`
+    - The current config
+   - The returned value (an arbitrary JSON value) is then used as the new config for the flow script.
+- A flow script can also export a `tick` function
+  - This function is called at a regular pace with the current time and config.
+  - The flow script can then return zero, one or many transformed messages
+  - By sharing an internal state between the `process` and `tick` functions,
+    the flow script can implement aggregations over a time window.
+    When messages are received they are pushed by the `process` function into that state
+    and the final outcome is extracted by the `tick` function at the end of the time window.
+
+## First release
+
+While the POC provides a generic mapper that is fully independent of the legacy mappers,
+the plan is not to abandon the latter in favor of the former
+but to revisit the legacy mappers to include the ability for users to add their own mapping rules.
+
+To be lovable, the first release of an extensible mapper should at least:
+
+- be a drop-in replacement of the current mapper (for c8y, aws, az or collect)
+- feature the ability to customize MEA processing by combining builtin flow steps with user-provided functions written in JavaScript
+- provide tools to create, test, monitor and debug steps and flows
+- be stable enough that user-defined flow scripts will still work without changes with future releases.
+
+To keep things simple for the first release, the following questions are deferred:
+
+- Could a generic mapper let users define bridge rules as well as message transformation flows?
+- Does it make sense to run such a mapper on child-devices?
+- Could a flow send HTTP messages? Or could a flow step tell the runtime to send messages over HTTP?
+- How to handle binary payloads on the MQTT bus? 
+- Could operations be managed is a similar way with user-provided functions to transform commands?
+- To handle operations, would the plugins be expanded to do more complex things like HTTP calls, file-system interactions, etc.? 
+- What are the pros and cons to persist flow step states?
+- Split a flow, forwarding transformed messages to different flows for further processing
+
+### API
+
+The POC expects the flow scripts to implement a bunch of functions. This gives a quite expressive interface
+(filtering, mapping, splitting, dynamic configuration, aggregation over time windows), but at the cost of some complexity.
+
+- `process(t: Timestamp, msg: Message, config: Json) -> Vec<Message>`
```

Comment:
```suggestion
- `process(t: Timestamp, msg: Message, context: Context) -> Vec<Message>`
```
Since we have plans to add context in the future anyway, how about adding it to the method signature in the initial version itself and expose the `Config` via this context object as `context.config`? We can keep adding more fields to the context like `context.device`, `context.tedgeConfig` etc in future as well.